File: skills/pudu-openapi-skill/scripts/pudu-request.py
# -*- coding: utf-8 -*-
"""
Pudu OpenAPI 请求工具

能力：
- 支持 GET / POST 的 HMAC-SHA1 签名请求
- 自动解析环境变量中的凭证和集群
- 统一返回解析后的结果，便于上层直接判断成功或失败

默认读取以下环境变量：
- PUDU_API_APP_KEY
- PUDU_API_APP_SECRET
- PUDU_API_CLUSTER（cn / sea / de / us）
"""

import os
import base64
import datetime
import hashlib
import hmac
import json
import re
from urllib.parse import quote, unquote
import requests

CLUSTERS = {
    "cn": {
        "name": "国内生产节点",
        "hostname": "open-platform.pudutech.com",
    },
    "sea": {
        "name": "海外（日韩新加坡）生产节点",
        "hostname": "css-open-platform.pudutech.com",
    },
    "de": {
        "name": "德国生产节点",
        "hostname": "csg-open-platform.pudutech.com",
    },
    "us": {
        "name": "美国生产节点",
        "hostname": "csu-open-platform.pudutech.com",
    },
}

# ─── 通用工具 ─────────────────────────────────────────────────────────────────

def mask_secret(value: str) -> str:
    """对敏感值做脱敏展示，避免日志或报错中回显完整凭证"""
    if not value:
        return ""
    if len(value) <= 6:
        return f"{value[:1]}***"
    return f"{value[:3]}***{value[-3:]}"

def is_mac_address(value) -> bool:
    """判断字符串是否是常见 MAC 地址格式，例如：14:80:CC:89:27:A6"""
    if not isinstance(value, str):
        return False
    return bool(re.match(r"^[0-9A-Fa-f]{2}(:[0-9A-Fa-f]{2}){5}$", value))

def validate_sn_field(params):
    """当接口需要 sn 时，防止误把 MAC 地址当作 sn 传入。"""
    if not isinstance(params, dict):
        return
    sn = params.get("sn")
    if is_mac_address(sn):
        raise ValueError(
            f"检测到 sn 的值看起来像 MAC 地址（{sn}）。请提供机器人 SN，而不是 MAC 地址。"
        )

def resolve_credentials(api_app_key=None, api_app_secret=None):
    """从环境变量和显式参数中解析凭证。显式参数优先，其次读取环境变量。"""
    resolved_key = api_app_key or os.environ.get("PUDU_API_APP_KEY")
    resolved_secret = api_app_secret or os.environ.get("PUDU_API_APP_SECRET")
    
    if not resolved_key:
        raise ValueError("ApiAppKey 缺失。请直接提供 api_app_key，或设置环境变量 PUDU_API_APP_KEY。")
    if not resolved_secret:
        raise ValueError("ApiAppSecret 缺失。请直接提供 api_app_secret，或设置环境变量 PUDU_API_APP_SECRET。")
        
    return resolved_key, resolved_secret

def resolve_cluster(hostname=None, cluster=None):
    """
    解析集群信息。
    - 优先使用显式传入的 hostname
    - 否则使用 cluster 或环境变量 PUDU_API_CLUSTER 映射 hostname
    """
    if hostname:
        cluster_name = CLUSTERS.get(cluster, {}).get("name", "自定义集群") if cluster else "自定义集群"
        return {
            "cluster": cluster or "custom",
            "clusterName": cluster_name,
            "hostname": hostname
        }
        
    cluster_key = cluster or os.environ.get("PUDU_API_CLUSTER")
    if not cluster_key:
        raise ValueError("PUDU_API_CLUSTER 缺失。请设置环境变量 PUDU_API_CLUSTER，或显式传入 cluster/hostname。")
        
    cluster_config = CLUSTERS.get(cluster_key)
    if not cluster_config:
        raise ValueError(f"无效的集群值：{cluster_key}。可选值仅支持 cn / sea / de / us。")
        
    return {
        "cluster": cluster_key,
        "clusterName": cluster_config["name"],
        "hostname": cluster_config["hostname"]
    }

# ─── 签名工具 ─────────────────────────────────────────────────────────────────

def compute_content_md5(body_str: str) -> str:
    """计算请求体的 Content-MD5（POST 专用）。规则：MD5(bodyStr) → hex → base64"""
    if not body_str:
        return ""
    md5_hex = hashlib.md5(body_str.encode("utf-8")).hexdigest()
    return base64.b64encode(md5_hex.encode("utf-8")).decode("utf-8")

def build_query_string(params: dict) -> str:
    """将参数对象按字典序拼接为 query string（GET 专用）"""
    parts = []
    for key in sorted(params.keys()):
        value = params[key]
        if isinstance(value, list):
            if value:
                encoded_values = [quote(str(v), safe='') for v in value]
                parts.append(f"{key}={','.join(encoded_values)}")
            else:
                parts.append(str(key))
        elif value is None or value == "":
            parts.append(str(key))
        else:
            parts.append(f"{key}={quote(str(value), safe='')}")
    return "&".join(parts)

def build_signing_string(method: str, sign_path: str, date_time: str, content_md5: str) -> str:
    """构造 HMAC 签名字符串"""
    return f"x-date: {date_time}\n{method}\napplication/json\napplication/json\n{content_md5}\n{sign_path}"

# ─── 核心请求方法 ──────────────────────────────────────────────────────────────

def request(
    path: str,
    method: str,
    params=None,
    hostname=None,
    cluster=None,
    api_app_key=None,
    api_app_secret=None,
    port=443,
    timeout_ms=15000
):
    """
    发起已签名的 HTTPS 请求（GET / POST）
    
    :param path: 接口路径（以 / 开头）
    :param method: 请求方法，"GET" 或 "POST"
    :param params: GET 查询参数 或 POST 请求体字段
    :param hostname: 请求 host（不含协议，例如 open-platform.pudutech.com）
    :param cluster: 集群标识（cn / sea / de / us），未传时读取 PUDU_API_CLUSTER
    :param api_app_key: 应用 ApiAppKey，未传时读取 PUDU_API_APP_KEY
    :param api_app_secret: 应用 ApiAppSecret，未传时读取 PUDU_API_APP_SECRET
    :param port: 端口号（HTTPS 默认 443）
    :param timeout_ms: 请求超时时间(毫秒)
    :return: 包含响应信息的字典
    """
    if params is None:
        params = {}
        
    if not path or not path.startswith("/"):
        raise ValueError("path 缺失或格式错误。请传入以 / 开头的接口路径。")
        
    normalized_method = method.upper()
    if normalized_method not in ["GET", "POST"]:
        raise ValueError(f"不支持的请求方法：{method}。当前仅支持 GET / POST。")
        
    api_key, api_secret = resolve_credentials(api_app_key, api_app_secret)
    cluster_info = resolve_cluster(hostname, cluster)
    is_get = normalized_method == "GET"
    
    validate_sn_field(params)
    
    # GET：参数进 query string；POST：参数序列化为 JSON body
    query_str = build_query_string(params) if is_get else ""
    body_str = "" if is_get else json.dumps(params, separators=(',', ':'), ensure_ascii=False)
    md5 = "" if is_get else compute_content_md5(body_str)
    
    # 请求路径（GET 附带 encoded query string）
    full_path = f"{path}?{query_str}" if query_str else path
    
    # 签名路径：剔除环境信息 (例如 /release, /test, /prepub)
    sign_path_base = path
    if sign_path_base.startswith(("/release", "/test", "/prepub")):
        parts = sign_path_base[1:].split("/", 1)
        if len(parts) > 1:
            sign_path_base = "/" + parts[1]
        else:
            sign_path_base = "/"
    sign_path_base = sign_path_base if sign_path_base else "/"
    
    # 签名路径：GET 使用 decoded query string，避免二次编码导致签名不匹配
    sign_path = f"{sign_path_base}?{unquote(query_str)}" if query_str else sign_path_base
    
    # 构造时间
    date_time = datetime.datetime.utcnow().strftime("%a, %d %b %Y %H:%M:%S GMT")
    
    # 构造签名
    signing_str = build_signing_string(normalized_method, sign_path, date_time, md5)
    signature = hmac.new(api_secret.encode("utf-8"), msg=signing_str.encode("utf-8"), digestmod=hashlib.sha1).digest()
    signature_b64 = base64.b64encode(signature).decode("utf-8")
    authorization = f'hmac id="{api_key}", algorithm="hmac-sha1", headers="x-date", signature="{signature_b64}"'
    
    url = f"https://{cluster_info['hostname']}:{port}{full_path}"
    headers = {
        "Accept": "application/json",
        "Content-Type": "application/json",
        "Content-MD5": md5,
        "x-date": date_time,
        "Authorization": authorization
    }
    
    try:
        if is_get:
            res = requests.get(url, headers=headers, timeout=timeout_ms / 1000.0)
        else:
            res = requests.post(url, headers=headers, data=body_str.encode("utf-8"), timeout=timeout_ms / 1000.0)
            
        status_code = res.status_code
        try:
            res_json = res.json()
        except ValueError:
            res_json = None
            
        message = res_json.get("message", "") if isinstance(res_json, dict) else ""
        ok = status_code == 200 and message == "SUCCESS"
        
        error_message = ""
        if not ok:
            if isinstance(res_json, dict):
                error_message = res_json.get("message") or res_json.get("error") or res_json.get("msg") or ""
            if not error_message:
                error_message = res.text or f"HTTP {status_code} 请求失败"
                
        return {
            "ok": ok,
            "status": status_code,
            "message": message,
            "cluster": cluster_info["cluster"],
            "clusterName": cluster_info["clusterName"],
            "hostname": cluster_info["hostname"],
            "method": normalized_method,
            "path": full_path,
            "json": res_json,
            "body": res.text,
            "traceId": res_json.get("trace_id", "") if isinstance(res_json, dict) else "",
            "errorMessage": error_message
        }
    except Exception as e:
        return {
            "ok": False,
            "status": 0,
            "message": str(e),
            "cluster": cluster_info["cluster"] if 'cluster_info' in locals() else "",
            "clusterName": cluster_info["clusterName"] if 'cluster_info' in locals() else "",
            "hostname": cluster_info["hostname"] if 'cluster_info' in locals() else "",
            "method": normalized_method if 'normalized_method' in locals() else method,
            "path": full_path if 'full_path' in locals() else path,
            "json": None,
            "body": "",
            "traceId": "",
            "errorMessage": str(e)
        }

# ─── 示例用法（直接执行时运行）────────────────────────────────────────────────

if __name__ == "__main__":
    try:
        api_key, api_secret = resolve_credentials()
        cluster_info = resolve_cluster()
        
        print("当前配置：")
        print(f"- ApiAppKey: {mask_secret(api_key)}")
        print(f"- ApiAppSecret: {mask_secret(api_secret)}")
        print(f"- 集群: {cluster_info['clusterName']} ({cluster_info['hostname']})")
        print("")
        print("请在其他脚本中按需调用 request()，避免直接执行本文件时误发真实请求。")
    except ValueError as e:
        print(f"配置检查失败：{str(e)}")

File: skills/pudu-openapi-skill/scripts/pudu-request.js
/**
 * Pudu OpenAPI 请求工具
 *
 * 能力：
 * - 支持 GET / POST 的 HMAC-SHA1 签名请求
 * - 自动解析环境变量中的凭证和集群
 * - 统一返回解析后的结果，便于上层直接判断成功或失败
 *
 * 默认读取以下环境变量：
 * - PUDU_API_APP_KEY
 * - PUDU_API_APP_SECRET
 * - PUDU_API_CLUSTER（cn / sea / de / us）
 */

"use strict";

const https = require("https");
const crypto = require("crypto");

const CLUSTERS = {
  cn: {
    name: "国内生产节点",
    hostname: "open-platform.pudutech.com",
  },
  sea: {
    name: "海外（日韩新加坡）生产节点",
    hostname: "css-open-platform.pudutech.com",
  },
  de: {
    name: "德国生产节点",
    hostname: "csg-open-platform.pudutech.com",
  },
  us: {
    name: "美国生产节点",
    hostname: "csu-open-platform.pudutech.com",
  },
};

// ─── 通用工具 ─────────────────────────────────────────────────────────────────

/**
 * 对敏感值做脱敏展示，避免日志或报错中回显完整凭证
 */
function maskSecret(value) {
  if (!value) {
    return "";
  }

  if (value.length <= 6) {
    return `${value.slice(0, 1)}***`;
  }

  return `${value.slice(0, 3)}***${value.slice(-3)}`;
}

/**
 * 安全解析 JSON。解析失败时返回 null，由上层决定如何展示原始文本。
 */
function parseJsonSafely(text) {
  if (!text) {
    return null;
  }

  try {
    return JSON.parse(text);
  } catch {
    return null;
  }
}

/**
 * 统一检查必填字段，缺失时抛出可读错误。
 */
function assertRequired(value, fieldName, hint) {
  if (value) {
    return;
  }

  const error = new Error(`${fieldName} 缺失。${hint}`);
  error.code = "PUDU_CONFIG_MISSING";
  throw error;
}

/**
 * 判断字符串是否是常见 MAC 地址格式，例如：14:80:CC:89:27:A6
 */
function isMacAddress(value) {
  return typeof value === "string" && /^[0-9A-Fa-f]{2}(:[0-9A-Fa-f]{2}){5}$/.test(value);
}

/**
 * 当接口需要 sn 时，防止误把 MAC 地址当作 sn 传入。
 */
function validateSnField(params) {
  if (!params || typeof params !== "object") {
    return;
  }

  if (isMacAddress(params.sn)) {
    const error = new Error(
      `检测到 sn 的值看起来像 MAC 地址（${params.sn}）。请提供机器人 SN，而不是 MAC 地址。`
    );
    error.code = "PUDU_SN_LOOKS_LIKE_MAC";
    throw error;
  }
}

/**
 * 从环境变量和显式参数中解析凭证。
 * 显式参数优先，其次读取环境变量。
 */
function resolveCredentials({ apiAppKey, apiAppSecret } = {}) {
  const resolvedAppKey = apiAppKey || process.env.PUDU_API_APP_KEY;
  const resolvedAppSecret = apiAppSecret || process.env.PUDU_API_APP_SECRET;

  assertRequired(
    resolvedAppKey,
    "ApiAppKey",
    "请直接提供 apiAppKey，或设置环境变量 PUDU_API_APP_KEY。"
  );
  assertRequired(
    resolvedAppSecret,
    "ApiAppSecret",
    "请直接提供 apiAppSecret，或设置环境变量 PUDU_API_APP_SECRET。"
  );

  return {
    apiAppKey: resolvedAppKey,
    apiAppSecret: resolvedAppSecret,
  };
}

/**
 * 解析集群信息。
 * - 优先使用显式传入的 hostname
 * - 否则使用 cluster 或环境变量 PUDU_API_CLUSTER 映射 hostname
 */
function resolveCluster({ hostname, cluster } = {}) {
  if (hostname) {
    return {
      cluster: cluster || "custom",
      clusterName: cluster && CLUSTERS[cluster] ? CLUSTERS[cluster].name : "自定义集群",
      hostname,
    };
  }

  const clusterKey = cluster || process.env.PUDU_API_CLUSTER;

  assertRequired(
    clusterKey,
    "PUDU_API_CLUSTER",
    "请设置环境变量 PUDU_API_CLUSTER，或显式传入 cluster/hostname。"
  );

  const clusterConfig = CLUSTERS[clusterKey];
  if (!clusterConfig) {
    const error = new Error(
      `无效的集群值：${clusterKey}。可选值仅支持 cn / sea / de / us。`
    );
    error.code = "PUDU_CLUSTER_INVALID";
    throw error;
  }

  return {
    cluster: clusterKey,
    clusterName: clusterConfig.name,
    hostname: clusterConfig.hostname,
  };
}

// ─── 签名工具 ─────────────────────────────────────────────────────────────────

/**
 * 计算 HMAC-SHA1 签名并返回 base64 字符串
 */
function hmacSha1(secret, message) {
  return crypto.createHmac("sha1", secret).update(message, "utf8").digest("base64");
}

/**
 * 计算请求体的 Content-MD5（POST 专用）
 * 规则：MD5(bodyStr) → hex → base64
 */
function computeContentMD5(bodyStr) {
  const hex = crypto.createHash("md5").update(bodyStr, "utf8").digest("hex");
  return Buffer.from(hex).toString("base64");
}

/**
 * 将参数对象按字典序拼接为 query string（GET 专用）
 *
 * 规则：
 *   - 值先 encodeURIComponent 编码
 *   - 数组用逗号分隔
 *   - 空值（null / undefined / ""）只保留 key
 */
function buildQueryString(params) {
  const parts = [];

  for (const key of Object.keys(params).sort()) {
    const value = params[key];

    if (Array.isArray(value)) {
      parts.push(value.length ? `${key}=${value.map(encodeURIComponent).join(",")}` : key);
    } else if (value === null || value === undefined || value === "") {
      parts.push(key);
    } else {
      parts.push(`${key}=${encodeURIComponent(value)}`);
    }
  }

  return parts.join("&");
}

/**
 * 构造 HMAC 签名字符串
 *
 * 格式（各行以 \n 分隔）：
 *   x-date: {dateTime}
 *   {METHOD}
 *   application/json        ← Accept
 *   application/json        ← Content-Type
 *   {contentMD5}
 *   {signPath}              ← GET 用 decoded path?query；POST 用 path
 */
function buildSigningString({ method, signPath, dateTime, contentMD5 }) {
  return [
    `x-date: ${dateTime}`,
    method,
    "application/json",
    "application/json",
    contentMD5,
    signPath,
  ].join("\n");
}

// ─── 核心请求方法 ──────────────────────────────────────────────────────────────

/**
 * 发起已签名的 HTTPS 请求（GET / POST）
 *
 * @param {object} opts
 * @param {string} [opts.hostname]    请求 host（不含协议，例如 open-platform.pudutech.com）
 * @param {string} [opts.cluster]     集群标识（cn / sea / de / us），未传时读取 PUDU_API_CLUSTER
 * @param {string} opts.path          接口路径（以 / 开头）
 * @param {string} opts.method        请求方法，"GET" 或 "POST"
 * @param {object} [opts.params={}]   GET 查询参数 或 POST 请求体字段
 * @param {string} [opts.apiAppKey]   应用 ApiAppKey，未传时读取 PUDU_API_APP_KEY
 * @param {string} [opts.apiAppSecret] 应用 ApiAppSecret，未传时读取 PUDU_API_APP_SECRET
 * @param {number} [opts.port=443]    端口号（HTTPS 默认 443）
 * @param {number} [opts.timeoutMs=15000] 请求超时时间
 * @returns {Promise<object>}
 */
function request({
  hostname,
  cluster,
  path,
  method,
  params = {},
  apiAppKey,
  apiAppSecret,
  port = 443,
  timeoutMs = 15000,
}) {
  return new Promise((resolve, reject) => {
    assertRequired(path, "path", "请传入以 / 开头的接口路径。");
    assertRequired(method, "method", "请传入 GET 或 POST。");

    const normalizedMethod = method.toUpperCase();
    if (!["GET", "POST"].includes(normalizedMethod)) {
      const error = new Error(`不支持的请求方法：${method}。当前仅支持 GET / POST。`);
      error.code = "PUDU_METHOD_UNSUPPORTED";
      reject(error);
      return;
    }

    const credentials = resolveCredentials({ apiAppKey, apiAppSecret });
    const clusterInfo = resolveCluster({ hostname, cluster });
    const isGet = normalizedMethod === "GET";
    const dateTime = new Date().toUTCString();

    validateSnField(params);

    // GET：参数进 query string；POST：参数序列化为 JSON body
    const queryStr = isGet ? buildQueryString(params) : "";
    const bodyStr = isGet ? "" : JSON.stringify(params);
    const md5 = isGet ? "" : computeContentMD5(bodyStr);

    // 请求路径（GET 附带 encoded query string）
    const fullPath = queryStr ? `${path}?${queryStr}` : path;

    // 签名路径：GET 使用 decoded query string，避免二次编码导致签名不匹配
    const signPath = queryStr ? `${path}?${decodeURIComponent(queryStr)}` : path;

    // 构造签名
    const signingStr = buildSigningString({
      method: normalizedMethod,
      signPath,
      dateTime,
      contentMD5: md5,
    });
    const signature = hmacSha1(credentials.apiAppSecret, signingStr);
    const authorization = `hmac id="${credentials.apiAppKey}", algorithm="hmac-sha1", headers="x-date", signature="${signature}"`;

    const options = {
      hostname: clusterInfo.hostname,
      port,
      path: fullPath,
      method: normalizedMethod,
      headers: {
        Accept: "application/json",
        "Content-Type": "application/json",
        "Content-MD5": md5,
        "Content-Length": Buffer.byteLength(bodyStr),
        "x-date": dateTime,
        Authorization: authorization,
      },
    };

    const req = https.request(options, (res) => {
      let body = "";
      res.on("data", (chunk) => (body += chunk));
      res.on("end", () => {
        const json = parseJsonSafely(body);
        const message = json && typeof json.message === "string" ? json.message : "";
        const ok = res.statusCode === 200 && message === "SUCCESS";
        const errorMessage =
          ok
            ? ""
            : (json && (json.message || json.error || json.msg)) ||
              body ||
              `HTTP ${res.statusCode} 请求失败`;

        resolve({
          ok,
          status: res.statusCode,
          message,
          cluster: clusterInfo.cluster,
          clusterName: clusterInfo.clusterName,
          hostname: clusterInfo.hostname,
          method: normalizedMethod,
          path: fullPath,
          json,
          body,
          traceId: json && json.trace_id ? json.trace_id : "",
          errorMessage,
        });
      });
    });

    req.setTimeout(timeoutMs, () => {
      req.destroy(new Error(`请求超时：${timeoutMs}ms`));
    });
    req.on("error", reject);
    req.write(bodyStr);
    req.end();
  });
}

module.exports = {
  CLUSTERS,
  isMacAddress,
  maskSecret,
  resolveCredentials,
  resolveCluster,
  request,
};

// ─── 示例用法（直接执行时运行）────────────────────────────────────────────────

if (require.main === module) {
  try {
    const credentials = resolveCredentials();
    const clusterInfo = resolveCluster();

    console.log("当前配置：");
    console.log(`- ApiAppKey: ${maskSecret(credentials.apiAppKey)}`);
    console.log(`- ApiAppSecret: ${maskSecret(credentials.apiAppSecret)}`);
    console.log(`- 集群: ${clusterInfo.clusterName} (${clusterInfo.hostname})`);
    console.log("");
    console.log("请在其他脚本中按需调用 request()，避免直接执行本文件时误发真实请求。");
  } catch (error) {
    console.error(`配置检查失败：${error.message}`);
  }
}
